### Pivots and their influence on solution spaces

Up until now, we have witnessed multiple instances that demonstrate how the reduced row echelon matrix provides a convenient way to describe the solution space of a linear system. In this section, we will utilize this comprehension to draw some general conclusions about how specific characteristics of the reduced row echelon matrix correspond to the properties of the solution space.

#### Pivot positions

A pivot position within matrix $A$ refers to the location of a leading element in the reduced row echelon form of matrix $A$.

It is important to recall that within a reduced row echelon matrix, a leading entry is defined as the first nonzero entry encountered from the left in each row of the matrix.

##### Example: Pivots in a matrix

Here's an example of a matrix in row reduced echelon form with its pivot positions indicated:

$$
\begin{bmatrix}
1 & 0 & 2 & 0 \\
0 & 1 & -3 & 0 \\
0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 \\
\end{bmatrix}
$$

In this matrix, the pivot positions are located at the positions where the leading entries are found. the pivot positions are indicated by the bold numbers below:

$$
\begin{bmatrix}
\mathbf{1} & 0 & 2 & 0 \\
0 & \mathbf{1} & -3 & 0 \\
0 & 0 & 0 & \mathbf{1} \\
0 & 0 & 0 & 0 \\
\end{bmatrix}
$$

These pivot positions play a crucial role in determining the solution space and the properties of the linear system associated with the matrix.

#### Number of solutions

As we have explored solution spaces of linear systems, we have often inquired about the presence of infinitely many solutions, a single solution, or no solutions. To delve deeper into this matter, we will now divide this question into two distinct questions:

1. **Existence**: Is there a solution to the linear system? If so, we say that the system is consistent; if not, we say it is inconsistent.
2. **Uniqueness**: If the linear system is consistent, is the solution unique or are there infinitely many solutions?

We will examine how the positioning of the pivots impacts the responses to these two inquiries. We commence by addressing the first question regarding the existence of solutions.

#### Existence

A linear system is considered **inconsistent** if and only if there exists a pivot position in the rightmost column of the corresponding augmented matrix.

##### Example: An inconsistent system

Here's an example of an inconsistent linear system with four variables:

$$
\begin{align*}
2x + 3y - z + 4w &= 10 \\
x - 2y + 3z - 5w &= -2 \\
3x - y + 2z - 6w &= 5 \\
4x + 6y - 2z + 8w &= 3 \\
\end{align*}
$$

We can represent this system using an augmented matrix:

$$
\begin{bmatrix}
2 & 3 & -1 & 4 & 10 \\
1 & -2 & 3 & -5 & -2 \\
3 & -1 & 2 & -6 & 5 \\
4 & 2 & -5 & 8 & 3 \\
\end{bmatrix}
$$

We can use SymPy to turn this matrix to RREF

In [4]:
from sympy import Matrix

M = Matrix([
    [2, 3, -1, 4, 10],
    [1, -2, 3, -5, -2],
    [3, -1, 2, -6, 5],
    [4, 6, -2, 8, 3]
])

M_rref, pivots = M.rref()

M_rref

Matrix([
[1, 0, 0, -3/2, 0],
[0, 1, 0,  5/2, 0],
[0, 0, 1,  1/2, 0],
[0, 0, 0,    0, 1]])

After row-reducing the augmented matrix to its reduced row echelon form, we obtain:

$$
\begin{bmatrix}
1 & 0 & 0 & -\frac{3}{2} & 1 \\
0 & 1 & 0 & \frac{5}{2} & 0 \\
0 & 0 & 1 & \frac{1}{2} & 0 \\
0 & 0 & 0 & 0 & 1 \\
\end{bmatrix}
$$

Since there is a pivot position in the rightmost column of the reduced row echelon form (in the forth row), the linear system is inconsistent. This implies that there are no solutions that satisfy all four equations simultaneously.

> If every row of the **coefficient matrix** contains a pivot position, then the system of linear equations associated with the matrix is considered to be consistent.

#### Uniqueness

Suppose that we consider a consistent linear system.

1. If there is a pivot position in each column of the **coefficient matrix**, then the system of linear equations has a single, **unique** solution.
2. If there exists a column in the **coefficient matrix** that does not have a pivot position, and the system is consistent, then the system of linear equations has an **infinite** number of solutions.
3. Pivot positions in the columns correspond to **basic variables**, while columns without pivot positions correspond to **free variables**.


##### Example

Consider the following linear system of equations:

$$
\begin{align*}
2x_1 - 3x_2 + 4x_3 + 5x_4 &= 10 \\
x_1 + 2x_2 - x_3 + 3x_4 &= 5 \\
3x_1 - x_2 + 2x_3 - 4x_4 &= -3 \\
4x_1 + x_2 + 3x_3 - 2x_4 &= 8 \\
\end{align*}
$$

This system of equations can be represented using the augmented matrix as:

$$
\left[
\begin{array}{cccc|c}
2 & -3 & 4 & 5 & 10 \\
1 & 2 & -1 & 3 & 5 \\
3 & -1 & 2 & -4 & -3 \\
4 & 1 & 3 & -2 & 8 \\
\end{array}
\right]
$$

Let's now find solutions to this system using SymPy and `rref` method.

In [2]:
from sympy import Matrix

M = Matrix([
    [2, -3, 4, 5, 10],
    [1, 2, -1, 3, 5],
    [3, -1, 2, -4, -3],
    [4, 1, 3, -2, 8]
])

M_rref, pivots = M.rref()

M_rref

Matrix([
[1, 0, 0, 0, -29/35],
[0, 1, 0, 0, 102/35],
[0, 0, 1, 0,   18/5],
[0, 0, 0, 1,    6/5]])

As we can see, this system has this unique solution:

$$
\begin{align}
x_1 &= -\frac{29}{35} \\
x_2 &= \frac{102}{35} \\
x_3 &= \frac{18}{5} \\
x_4 &= \frac{6}{5}
\end{align}
$$

Now consider this system of equations:

$$
\begin{align*}
x_1 + 2x_2 - 3x_3 + x_4 &= 4 \\
2x_1 + 4x_2 - 6x_3 + 2x_4 &= 8 \\
3x_1 + 6x_2 - 9x_3 + 3x_4 &= 12 \\
\end{align*}
$$

This system of equations can be represented using the augmented matrix as:

$$
\left[
\begin{array}{cccc|c}
1 & 2 & -3 & 1 & 4 \\
2 & 4 & -6 & 2 & 8 \\
3 & 6 & -9 & 3 & 12 \\
\end{array}
\right]
$$

Let's find solution space for this system as well.

In [3]:
from sympy import Matrix

M = Matrix([
    [1, 2, -3, 1, 4],
    [2, 4, -6, 2, 8],
    [3, 6, -9, 3, 12]
])

M_rref, pivots = M.rref()

M_rref

Matrix([
[1, 2, -3, 1, 4],
[0, 0,  0, 0, 0],
[0, 0,  0, 0, 0]])

The system is **consistent** because there is no pivot in the rightmost column of the corresponding augmented matrix. But the system does not have a unique solution, because the second, third, and fourth columns in coefficient matrix do not have a pivot.

$x_1$ is the basic variable, and $x_2$ , $x_3$, and $x_4$ are free variables. The solution space is:

$$
\begin{align}
x_1 + 2x_2 - 3x_3 + x_4 &= 4 \\
x_2, x_3, x_4 &\in \mathbb{R}
\end{align}
$$